jax/experimental/shard_map.py

```diff
@@ -985,19 +987,20 @@ def out_names_transform(out_names):
 @lu.transformation
 def _promote_scalar_residuals(*args, **kwargs):
   jaxpr, (out_pvals, out_consts, env) = yield args, kwargs
-  which_scalar = [isinstance(v.aval, core.ShapedArray) and not v.aval.shape
-                  for v in jaxpr.constvars]
-  out_consts_ = [jax.lax.broadcast(x, (1,)) if scalar else x
-                 for x, scalar in zip(out_consts, which_scalar)]
+  jaxpr, out_consts = _promote_scalar_residuals_jaxpr(jaxpr, out_consts)
+  yield jaxpr, (out_pvals, out_consts, env)
+
+def _promote_scalar_residuals_jaxpr(jaxpr, res):
+  which = [isinstance(v.aval, core.ShapedArray) and not v.aval.shape
+           for v in jaxpr.constvars]
+  res_ = [jax.lax.broadcast(x, (1,)) if s else x for x, s in zip(res, which)]
```

Comment:
Howcome we don't `broadcast` like this in `_add_singleton` (and then use it here too, in turn)?

Reply:
I think here `x` might be of Python builtin numeric type, so we can't count on it having a `reshape` method, whereas `_add_singleton` is currently called in places where `x` is guaranteed to have a `reshape` method. We could probably unify these but... probably not worth much!

Reply:
Yep, that makes sense.